File: 01458. Max Dot Product of Two Subsequences/solution.js
/**
 * Problem: 1458. Max Dot Product of Two Subsequences
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Calculate maximum dot product of two subsequences
 *
 * @param {number[]} nums1 - First integer array
 * @param {number[]} nums2 - Second integer array
 *
 * @returns {number} - Maximum dot product
 */
const maxDotProduct = (nums1, nums2) => {
    // Get lengths of both arrays
    const m = nums1.length
    const n = nums2.length

    // Ensure nums1 is the longer array for space optimization
    if (m < n) return maxDotProduct(nums2, nums1)

    // Initialize DP array with -Infinity
    const dp = new Array(n + 1).fill(-Infinity)

    // Iterate through each element of nums1
    for (let i = 0; i < m; i++) {
        // prev stores dp[j] from previous iteration of i
        let prev = 0

        // Iterate through each element of nums2
        for (let j = 0; j < n; j++) {
            // Store current dp[j+1] before updating
            const current = dp[j + 1]

            // Update dp[j+1] with three possibilities:
            // 1. Take product with current pair + max(prev, 0)
            // 2. Skip current element from nums2 (dp[j])
            // 3. Skip current element from nums1 (current dp[j+1])
            dp[j + 1] = Math.max(
                Math.max(prev, 0) + nums1[i] * nums2[j],
                dp[j],
                dp[j + 1]
            )

            // Update prev for next iteration
            prev = current
        }
    }

    // Return maximum dot product
    return dp[n]
}

File: 01458. Max Dot Product of Two Subsequences/solution.cpp
/**
 * Problem: 1458. Max Dot Product of Two Subsequences
 *
 * Difficulty: Hard
 *
 * Language: C++
 *
 * Performance: Runtime - 1 ms (Beats 100%)
 */

class Solution {
public:
    int maxDotProduct(vector<int>& nums1, vector<int>& nums2) {
        // Get lengths of both arrays
        int n = nums1.size();
        int m = nums2.size();

        // Constant for negative infinity initialization
        const int NEG = -1e9;

        // DP table: dp[i][j] = max dot product for nums1[0..i-1] and
        // nums2[0..j-1]
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, NEG));

        // Fill DP table
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                // Option 1: Take current pair and possibly extend previous
                // subsequence
                int takePair =
                    nums1[i - 1] * nums2[j - 1] + max(0, dp[i - 1][j - 1]);

                // Update dp[i][j] with maximum of three options:
                // 1. Take current pair (with optional extension)
                // 2. Skip current element from nums1
                // 3. Skip current element from nums2
                dp[i][j] = max({takePair, dp[i - 1][j], dp[i][j - 1]});
            }
        }

        // Return maximum dot product for complete arrays
        return dp[n][m];
    }
};
